### PR TITLE
feat: implement scm contributes

### PR DESCRIPTION
**Pull request by Ricbet**
### Types

- [x] 🎉 New Features

### <samp>🤖 Generated by Copilot at 5b91ede</samp>

*  Add support for custom components for the SCM input box from extension contributions ([link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-7e4610f55daeaaca635b72e1cba7d5548f4d5ad6d5e36466dc1b26052ca015c1R271-R319), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-21a4aa56a7c4522217936bd1b863db24976d36fd1f97fb62ce75fdb01d696cffR10), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-21a4aa56a7c4522217936bd1b863db24976d36fd1f97fb62ce75fdb01d696cffL29-R31), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-b042bb48a2e3674524d4a40fa3a875d719d02925e7ddeb0bdf1c14711f457ebbR1-R65), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-93988976f20baaff79ef4701631c72f4e0faec1b15142189c262dbc6bc50510bR51-R55), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-58d6d4cbf584245bfbfd30d022f8ef46f0da79f24bcafc0c4b77e7d69daf1a93R46-R68), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-58d6d4cbf584245bfbfd30d022f8ef46f0da79f24bcafc0c4b77e7d69daf1a93R124))
*  Add support for custom input props for the SCM input box from extension contributions ([link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-7e4610f55daeaaca635b72e1cba7d5548f4d5ad6d5e36466dc1b26052ca015c1R16), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-7e4610f55daeaaca635b72e1cba7d5548f4d5ad6d5e36466dc1b26052ca015c1R83-R85), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-b042bb48a2e3674524d4a40fa3a875d719d02925e7ddeb0bdf1c14711f457ebbR1-R65), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-93988976f20baaff79ef4701631c72f4e0faec1b15142189c262dbc6bc50510bR1), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-93988976f20baaff79ef4701631c72f4e0faec1b15142189c262dbc6bc50510bR51-R55), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-58d6d4cbf584245bfbfd30d022f8ef46f0da79f24bcafc0c4b77e7d69daf1a93R35), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-58d6d4cbf584245bfbfd30d022f8ef46f0da79f24bcafc0c4b77e7d69daf1a93R46-R68), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-58d6d4cbf584245bfbfd30d022f8ef46f0da79f24bcafc0c4b77e7d69daf1a93R124), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-38b9602bcd1e5fa40c0e82b5f43537711b3a1c2302ab69493fa570d814bc3861R4), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-38b9602bcd1e5fa40c0e82b5f43537711b3a1c2302ab69493fa570d814bc3861R51-R69), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-38b9602bcd1e5fa40c0e82b5f43537711b3a1c2302ab69493fa570d814bc3861R206-R217), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-527faf067be61f7526b031487475180346065c5b81e644b85181bb33428e7ec6R1), [link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-527faf067be61f7526b031487475180346065c5b81e644b85181bb33428e7ec6R39-R43))
*  Add schema and default snippet for the `scm` section of the extension contributions ([link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-b042bb48a2e3674524d4a40fa3a875d719d02925e7ddeb0bdf1c14711f457ebbR1-R65))
*  Add Chinese translations for the `scm` section and its properties ([link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-77fd021913ae53d95584cf31e7a51658aca21ae8f865adf8d0e97b8205d04ac3R188-R192))
*  Add an empty line to the end of the file `packages/components/src/input/Input.tsx` to follow the code style convention ([link](https://github.com/opensumi/core/pull/2867/files?diff=unified&w=0#diff-bb1200d7ed8bb83c25174fe4873865d715373eb69ee26d66eae4779bad778611R189))

### <samp>🤖 Generated by Copilot at 5b91ede</samp>

This pull request adds support for extensions to contribute custom components for the SCM input box in the browser. It defines a new `SCMContributionPoint` class that handles the `scm` section of the extension contributions and calls the `SCMService` to set the input props. It also enhances the `SCMInput` and `SCMResourceInput` classes to support more input features and options. It updates the `IInputBaseProps` interface and the `Input` component to support the new input props. It also adds the Chinese translations for the `scm` section.

### CODE REVIEW

**Comment by pipiiiiii** [comment]
加个截图吧

**Comment by Ricbet** [comment]
> 加个截图吧

没有截图，最终效果其实就是 scm input 上的那些 props 属性

File: packages/i18n/src/common/contributes/zh-CN.lang.ts [hunk 1]
```diff
@@ -185,4 +185,9 @@ inline 模式 showTitle 会失效, 只显示 icon`,
     '(可选) 用于表示 UI 中的子菜单的图标。文件路径、具有深色和浅色主题的文件路径的对象，或者主题图标引用(如 "$(zap)")',
   'sumiContributes.submenus.icon.light': '使用浅色主题时的图标路径',
   'sumiContributes.submenus.icon.dark': '使用深色主题时的图标路径',
+
+  // scm
+  'sumiContributes.scm': '提供自定义的 scm 控件',
```
Comment (erha19) [inline packages/i18n/src/common/contributes/zh-CN.lang.ts:190]: 这里的 SCM 是缩写，用大写的吧

File: packages/scm/src/browser/components/scm-resource-input.tsx [hunk 2]
```diff
@@ -39,8 +41,44 @@ export const SCMResourceInput: FC<{
...
+    (props: IInputBaseProps) => {
+      const { addonAfter, addonBefore } = props;
+      const AFC = addonAfter;
+      const ABC = addonBefore;
+
+      setInputProps({
+        ...props,
+        ...(addonAfter
+          ? {
+              addonAfter: typeof AFC === 'function' ? <AFC /> : addonAfter,
```
Comment (erha19) [inline packages/scm/src/browser/components/scm-resource-input.tsx:54]: 可以用 `isFunction` 代替

File: packages/extension/src/browser/extension-view.service.ts [hunk 3]
```diff
@@ -264,6 +269,67 @@ export class ViewExtProcessService implements AbstractViewExtProcessService {
...
+
+        const appendComponent = (addon: string[], key: string) => {
+          addon.forEach((addonId: string) => {
+            const component = moduleExports[addonId];
+
+            if (!component) {
+              this.logger.error(`Can not find ${key} from extension ${extension.id}, id: ${component}`);
+            }
+
+            if (this.appConfig.useExperimentalShadowDom) {
```
Comment (erha19) [inline packages/extension/src/browser/extension-view.service.ts:292]: 这个没有最终拓展效果的示例图吗？

**Review by erha19 — CHANGES_REQUESTED** [review]

**Comment by Ricbet** [comment]
image
例如提供了 addonAfter 之后就会在 input 输入框右边渲染自定义的组件
@pipiiiiii @erha19

File: packages/extension/src/browser/extension-view.service.ts [hunk 4]
```diff
@@ -264,6 +269,67 @@ export class ViewExtProcessService implements AbstractViewExtProcessService {
...
+            } else {
+              repo.input.appendProps({
+                [key]: component,
+              });
+            }
+          });
+        };
+
+        if (addonBefore) {
+          appendComponent(addonBefore, 'addonBefore');
```
Comment (erha19) [inline packages/extension/src/browser/extension-view.service.ts:315]: `addonBefore` 和 `addonAfter` 是不是能统一到一个 SCMAddonPosition 之类的变量里面？

**Review by erha19 — APPROVED** [review]

**Pull request merged**